service/watcher/chartstatus/kubeconfig.go

```diff
@@ -0,0 +1,114 @@
+package chartstatus
+
+import (
+	"context"
+	"time"
+
+	"github.com/giantswarm/apiextensions/v3/pkg/clientset/versioned"
+	"github.com/giantswarm/app/v4/pkg/key"
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/errors/tenant"
+	"github.com/giantswarm/kubeconfig/v4"
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+
+	"github.com/giantswarm/app-operator/v3/pkg/label"
+)
+
+// getG8sClient returns a versioned clientset for the kubeconfig used by the
+// chart-operator app CR running in the same namespace as the operator.
+func (c *ChartStatusWatcher) getG8sClient(ctx context.Context) (versioned.Interface, error) {
+	lo := metav1.ListOptions{
+		LabelSelector: label.ChartOperatorAppSelector(c.uniqueApp),
+	}
+	apps, err := c.k8sClient.G8sClient().ApplicationV1alpha1().Apps(c.appNamespace).List(ctx, lo)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	if len(apps.Items) != 1 {
+		return nil, microerror.Maskf(executionFailedError, "expected 1 chart-operator app CR got %d", len(apps.Items))
+	}
+
+	// We have the chart-operator app CR. Now we get its kubeconfig.
+	app := apps.Items[0]
+
+	// App CR uses inCluster so we can reuse the existing client.
+	if key.InCluster(app) {
+		return c.k8sClient.G8sClient(), nil
+	}
+
+	var kubeConfig kubeconfig.Interface
+	{
+		c := kubeconfig.Config{
+			K8sClient: c.k8sClient.K8sClient(),
+			Logger:    c.logger,
+		}
+
+		kubeConfig, err = kubeconfig.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var restConfig *rest.Config
+	{
+		restConfig, err = kubeConfig.NewRESTConfigForApp(ctx, key.KubeConfigSecretName(app), key.KubeConfigSecretNamespace(app))
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var g8sClient versioned.Interface
+	{
+		c := rest.CopyConfig(restConfig)
+
+		g8sClient, err = versioned.NewForConfig(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	return g8sClient, nil
+}
+
+// waitForActiveKubeConfig gets a kubeconfig for the chart-operator app CR.
+// If the target cluster is remote then sometimes the connection will be down
+// so we list all chart CRs to confirm the connection is active.
+func (c *ChartStatusWatcher) waitForActiveKubeConfig(ctx context.Context) (versioned.Interface, error) {
```

Comment:
What do you think divde a function in here, first, we wait for active connection and get g8sClient later. Or we can adjust function name like `getAvailableG8sClient`.

Reply:
Yes good to split it and if InCluster is true I don't think we need to list the chart CRs to check the connection.